### PR TITLE
PeersResponse: always include connected peers

### PR DESCRIPTION
**Pull request by saketh-are**
PeersRequest/PeersResponse are used to share information about peers in the network to which a node may connect.

Currently, PeersResponse contains a randomly selected subset of the responding node's PeerStore. However, the PeerStore includes unverified data reported by other nodes, potentially including fake information from malicious nodes.

This PR changes PeerResponse to always include the peers directly connected to the responding node. It ensures that some useful information is passed even when the responding node is under attack and flooded with fake peers.

### CODE REVIEW

**Review by pompon0 — COMMENTED** [review]
LGTM, however what I actually had in mind was to distinguish direct connections from PeerStore content on the proto level: https://cs.github.com/near/nearcore/blob/b602b52e26cb56f5f54564707ea62ca7f08f672f/chain/network/src/network_protocol/network.proto#L297 (i.e. add direct_peers field there). This way we could:
* explictly treat the direct_peers with more trust
* put a strict limit on the number of direct_peers field size (for example: at most 40 entries)
Also I suspect we might want to eventually add more data for direct connections: inbound/outbound, connection age, etc. , in which case we will need a separate field anyway. You don't have to apply these suggestions for now, but perhaps some of them might simplify your next PRs.

File: chain/network/src/network_protocol/mod.rs [hunk 1]
```diff
@@ -347,7 +347,7 @@ pub enum PeerMessage {
     SyncAccountsData(SyncAccountsData),
 
     PeersRequest,
-    PeersResponse(Vec<PeerInfo>),
+    PeersResponse(Vec<PeerInfo>, Vec<PeerInfo>),
```
Comment (pompon0) [inline chain/network/src/network_protocol/mod.rs:350]: please name these fields and document them

File: chain/network/src/network_protocol/network.proto [hunk 2]
```diff
@@ -288,14 +288,16 @@ message SyncAccountsData {
...
-// - peers directly connected to the receiver
-// - peers that the receiver transitively learned about from other peers.
+// See PeersResponse below for the response.
 message PeersRequest {}
 
-// Response to PeersRequest.
+// Response to PeersRequest. Includes
+//  - peers: peers drawn from the PeerStore of the responding node,
+//           which includes peers learned transitively from other peers
+//  - direct_peers: peers directly connected to the responding node
```
Comment (pompon0) [inline chain/network/src/network_protocol/network.proto:297]: move these doc-lines directly next to the field definitions.

Comment (pompon0) [inline chain/network/src/network_protocol/network.proto:297]: please put a constraint on the size of this field (and document it)

File: chain/network/src/network_protocol/tests.rs [hunk 3]
```diff
@@ -116,7 +116,10 @@ fn serialize_deserialize() -> anyhow::Result<()> {
         PeerMessage::SyncRoutingTable(data::make_routing_table(&mut rng)),
         PeerMessage::RequestUpdateNonce(data::make_partial_edge(&mut rng)),
         PeerMessage::PeersRequest,
-        PeerMessage::PeersResponse((0..5).map(|_| data::make_peer_info(&mut rng)).collect()),
+        PeerMessage::PeersResponse(
+            (0..5).map(|_| data::make_peer_info(&mut rng)).collect(),
+            vec![],
```
Comment (pompon0) [inline chain/network/src/network_protocol/tests.rs:121]: please add a todo to populate this field, once borsh support is dropped.

File: chain/network/src/peer/peer_actor.rs [hunk 4]
```diff
@@ -1117,22 +1117,29 @@ impl PeerActor {
...
                 let node_id = self.network_state.config.node_id();
                 self.network_state.peer_store.add_indirect_peers(
                     &self.clock,
                     peers.into_iter().filter(|peer_info| peer_info.id != node_id),
                 );
+                // Direct peers of the responding peer are still indirect peers for this node.
+                // However, we may treat them with more trust in the future.
+                self.network_state.peer_store.add_indirect_peers(
+                    &self.clock,
+                    direct_peers.into_iter().filter(|peer_info| peer_info.id != node_id),
```
Comment (pompon0) [inline chain/network/src/peer/peer_actor.rs:1141]: while you are at it, perhaps move this filtering condition inside peer_store?

**Review by pompon0 — APPROVED** [review]

File: chain/network/src/peer/peer_actor.rs (hunk 4 above)
Comment (saketh-are) [reply chain/network/src/peer/peer_actor.rs:1144]: As discussed offline we will leave this as it is because peer_store does not know the node_id.

Comment (saketh-are) [reply chain/network/src/peer/peer_actor.rs:1144]: As discussed offline, peer_store does not know the node_id. I agree that having the filtering here is a little unfortunate but I am leaving it as is for now.

File: chain/network/src/network_protocol/network.proto (hunk 2 above)
Comment (saketh-are) [reply chain/network/src/network_protocol/network.proto]: Looking into it we don't currently constrain the size of the PeersResponse on the receiving end, just on the sending: https://github.com/near/nearcore/blob/master/chain/network/src/config.rs#L121

I'll send a separate PR constraining both the sizes of `peers` and `direct_peers` on the receiving end, with a ban for misbehaving peers.

**Pull request merged**